### PR TITLE
Parent-child relationships

### LINKED ISSUE #1202: template stacks: track parent-child relationships
### What problem are you facing?

When a template stack controller creates some resources, we want to keep track of the relationship between the parent and the children, for a few reasons:

* Garbage collection
* Resource discovery (as a user)

### How could Crossplane help solve your problem?

We already have some separate work which has explored how to represent parent child relationships. The template stack controller should set the information necessary to describe the relationships based on the scheme we agreed on, which is owner references and certain labels, if I recall correctly.

Creating the labels and owner references for the direct children should be fairly easy, since the template stack controller is creating those. However, those children may create children too, and we want to be able to find those, so getting those indirect children tracked properly may be trickier.

### What does it look like when it's done?

* Resources applied to the cluster as the result of a resource engine's execution in the template stack controller have labels added to them to track the parent, in accordance with our [labeling scheme](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-stack-relationship-labels.md)
* This means that, for example, if a user creates a `MinimalGCP` CR, all resources rendered as a result have the labels
* Labels only need to go one level down; indirect children are out of scope

### PR DESCRIPTION
**Pull request by muvaf**
This commit introduces the parent-child relationships via labels as specified in https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-stack-relationship-labels.md

Fixes https://github.com/crossplaneio/crossplane/issues/1202

### CODE REVIEW

**Review by suskin — APPROVED** [review]
Nice - looks good to me! :squid:

One question - what testing has been done?

Approving in case it's already been tested : )

File: pkg/resource/api.go [hunk 1]
```diff
@@ -102,6 +106,58 @@ func (lo NamespacePatcher) Patch(cr ParentResource, list []ChildResource) ([]Chi
...
+		if cr.GetNamespace() != "" {
+			labels[stacks.LabelParentNamespace] = cr.GetNamespace()
+		}
+		labels[stacks.LabelParentName] = cr.GetName()
+		labels[stacks.LabelParentUID] = string(cr.GetUID())
+		gvk := cr.GetObjectKind().GroupVersionKind()
+		labels[stacks.LabelParentKind] = gvk.Kind
+		labels[stacks.LabelParentGroup] = gvk.Group
+		labels[stacks.LabelParentVersion] = gvk.Version
+		o.SetLabels(labels)
```
Comment (displague) [inline pkg/resource/api.go]: `stacks.ParentLabels` (crossplane) and `meta.AddLabels` (crossplane-runtime) could minimize the code needed here.

Comment (displague) [reply pkg/resource/api.go]: `meta.AddLabels(o, stacks.ParentLabels(cr))`

File: pkg/resource/api.go [hunk 2]
```diff
@@ -102,6 +106,58 @@ func (lo NamespacePatcher) Patch(cr ParentResource, list []ChildResource) ([]Chi
...
+func (lo LabelPropagator) Patch(cr ParentResource, list []ChildResource) ([]ChildResource, error) {
+	for _, o := range list {
+		labels := o.GetLabels()
+		if labels == nil {
+			labels = map[string]string{}
+		}
+		for key, val := range cr.GetLabels() {
+			labels[key] = val
+		}
+		o.SetLabels(labels)
```
Comment (displague) [inline pkg/resource/api.go]: `meta.AddLabels(o, cr.GetLabels())`

**Comment by muvaf** [comment]
@suskin Thanks! I've tested it with minimal gcp stack and it's working:
```
  labels:
    core.crossplane.io/parent-group: gcp.resourcepacks.crossplane.io
    core.crossplane.io/parent-kind: MinimalGCP
    core.crossplane.io/parent-name: test
    core.crossplane.io/parent-uid: d3f8fa83-e142-4fac-b207-85aa345d1031
    core.crossplane.io/parent-version: v1alpha1
```

**Pull request merged**